### ISSUE TITLE
salt-cloud ec2 driver missing "Deploy: False" option

### ISSUE THREAD

**Issue by johnccfm** [saltstack/salt#29085]
I'm trying to create Windows EC2 instances with salt-cloud using the **ec2** driver. I want to handle bootstrapping of the box myself, so I added:

    deploy: False
    script: None

to my profiles. However, after creating the device salt-cloud still attempts to SSH into the box and takes a very long time to time out.

Looking at the code briefly, it looks to me like the **aws** driver had the "deploy: False" option but it's not in the **ec2** driver, which I think is a bug.

**Comment by techhat** [comment]
Thanks for the report! This did work at one point with the EC2 driver, so this is a regression type bug. Could you please post the output of `salt-cloud --versions-report`?

**Comment by johnccfm** [comment]
@techhat I can't say that it ever actually did work with EC2. I think on balance, probably not. Sorry I can't be more certain.

```
johncc@saltmaster:/srv/salt/pillar/os$ salt-cloud --versions
Salt Version:
            Salt: 2015.8.1
 
Dependency Versions:
 Apache Libcloud: 0.15.1
          Jinja2: 2.7.2
        M2Crypto: Not Installed
            Mako: 0.9.1
          PyYAML: 3.10
           PyZMQ: 14.4.0
          Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
            RAET: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.0.4
            cffi: Not Installed
        cherrypy: Not Installed
        dateutil: 2.4.2
           gitdb: 0.5.4
       gitpython: 0.3.2 RC1
           ioflo: Not Installed
         libnacl: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.3.0
    mysql-python: 1.2.3
       pycparser: Not Installed
        pycrypto: 2.6.1
          pygit2: Not Installed
    python-gnupg: Not Installed
           smmap: 0.8.2
         timelib: Not Installed
 
System Versions:
            dist: Ubuntu 14.04 trusty
         machine: x86_64
         release: 3.13.0-49-generic
          system: Ubuntu 14.04 trusty
```

**Comment by techhat** [comment]
I have used it a number of times, though not recently, so I know that it worked at one point. And certainly, it needs to get fixed. Thanks again!

**Comment by pass-by-value** [comment]
@johnccfm 
Created #29576 for this. Added the `seed` option that controls whether the bootstrap script is run. Using the name `seed` since I've seen it in the lxc driver as well. Please let me know if I should change it to `deploy` (not sure what the standard name is for this) @techhat

**Comment by techhat** [comment]
@pass-by-value you need to look at `deploy_config` in `salt/utils/cloud.py`, in the `bootstrap()` function.

**Comment by pass-by-value** [comment]
@johnccfm 
I am able to have it bypass the bootstrap by setting `deploy: False` my cloud profile

```yaml
adi_ec2_testing:
  provider: my-ec2-us-east-public-ips
  image: ami-01c3bc64
  size: m3.medium
  ssh_username: salt
  deploy: False
```

This is seen in the output of the salt command

```yaml
adi_arch_test1:
    ----------
    Error:
        ----------
        No Deploy:
            'deploy' is not enabled. Not deploying.
    amiLaunchIndex:
        0
    architecture:
        x86_64
    blockDeviceMapping:
        ----------
        item:
            ----------
            deviceName:
                /dev/sda1
            ebs:
                ----------
                attachTime:
                    2015-12-09T19:18:20.000Z
                deleteOnTermination:
                    true
                status:
                    attaching
                volumeId:
                    sanitized
    clientToken:
        None
    dnsName:
        sanitized
    ebsOptimized:
        false
    groupSet:
        ----------
        item:
            ----------
            groupId:
                sanitized
            groupName:
                adi_salt_cloud
    hypervisor:
        xen
    imageId:
        ami-01c3bc64
    instanceId:
        sanitized
    instanceState:
        ----------
        code:
            0
        name:
            pending
    instanceType:
        m3.medium
    ipAddress:
        sanitized
    keyName:
        sanitized
    launchTime:
        2015-12-09T19:18:16.000Z
    monitoring:
        ----------
        state:
            disabled
    networkInterfaceSet:
        None
    placement:
        ----------
        availabilityZone:
            us-east-1b
        groupName:
            None
        tenancy:
            default
    privateDnsName:
        sanitized
    privateIpAddress:
        sanitized
    productCodes:
        None
    reason:
        None
    rootDeviceName:
        /dev/sda1
    rootDeviceType:
        ebs
    virtualizationType:
        hvm
```

Here are the particulars of my installation

```shell
root@localhost:~/salt# salt-cloud --versions-report
Salt Version:
            Salt: 2015.8.1

Dependency Versions:
 Apache Libcloud: Not Installed
          Jinja2: 2.7.3
        M2Crypto: 0.21.1
            Mako: Not Installed
          PyYAML: 3.11
           PyZMQ: 14.4.0
          Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
            RAET: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.0.5
            cffi: 0.8.6
        cherrypy: Not Installed
        dateutil: Not Installed
           gitdb: Not Installed
       gitpython: Not Installed
           ioflo: Not Installed
         libnacl: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.4.2
    mysql-python: Not Installed
       pycparser: 2.10
        pycrypto: 2.6.1
          pygit2: NotInstalled
    python-gnupg: 0.3.8
           smmap: Not Installed
         timelib: Not Installed

System Versions:
            dist: debian 8.1
         machine: x86_64
         release: 4.1.5-x86_64-linode61
          system: debian 8.1

root@localhost:~/salt#
```
 I should add that it will still try to ssh to the newly created instance, but it did bypass the bootstrap process.

```shell
[DEBUG   ] Terminal Command: /bin/sh -c ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPasswordAuthentication=no -oChallengeResponseAuthentication=no -oPubkeyAuthentication=yes -oKbdInteractiveAuthenti
cation=no -i /root/foo.pem -p 22 root@1.2.3.4 date
```

Could you please post your cloud profile? That way I can try it out to make sure I didn't miss anything.

**Comment by pass-by-value** [comment]
@johnccfm I'm wondering if you got a chance to try this out?

**Comment by johnccfm** [comment]
Sorry, I'm being slow. Do you mean that deploy: False already works in the current version? In that case, I was probably seeing the attempted SSH connection to the minion and wrongly interpreting that as an attempt to start deployment.

If so, then it's still a problem, just a slightly different one. SSH to a Windows host is never going to work. This means bringing up Windows boxes with salt-cloud involves waiting for a long, long timeout or manually killing the process. Is there any way to prevent the SSH attempt?

**Comment by techhat** [comment]
It should already be working. If Salt Cloud is still attempting an SSH connection, then it is definitely broken. @pass-by-value, could you please take another look?

If you specify a Windows installer, then Salt Cloud will attempt to use SMB and `winexe` to log into the box instead. But you want to bootstrap it yourself, so that doesn't help.

**Comment by pass-by-value** [comment]
Hi @techhat. It seems like the call to [wait_for_instance](https://github.com/saltstack/salt/blob/454efa168812bef7f2543994e9198ef239017ca3/salt/cloud/clouds/ec2.py#L2457-L2459) is dialing out to the vm via ssh. It happens regardless of whether `deploy` is set to True or False. I could add a new config option to the cloud profile to bypass that call. Do you think this is a good way to stop the SSH attempt?

Thanks for the clarification @johnccfm, sorry I didn't understand the query earlier!

**Comment by pass-by-value** [comment]
Created #29895 for this

**Comment by rallytime** [comment]
This is fixed at the HEAD of the 2015.8 branch and the fix will be available in the upcoming 2015.8.4 release.
